src/middleware.rs

```diff
@@ -0,0 +1,138 @@
+use std::marker::PhantomData;
+
+use Service;
+
+/// Often, many of the pieces needed for writing network applications
+/// can be reused across multiple services. The `Middleware` trait can
+/// be used to write reusable components that can be applied to very
+/// different kinds of services; for example, it can be applied to
+/// services operating on different protocols, and to both the client
+/// and server side of a network transaction.
+///
+/// # Timeouts
+///
+/// Take timeouts as an example:
+///
+/// ```rust,ignore
+/// use tokio::Service;
+/// use tokio::Middleware;
+/// use futures::Future;
+/// use std::time::Duration;
+///
+/// // Not yet implemented, but soon :)
+/// use tokio::timer::{Timer, Expired};
+///
+///
+/// pub struct Timeout {
+///     delay: Duration,
+///     timer: Timer,
+/// }
+///
+/// impl Timeout {
+///     fn timeout(&self) -> impl Future<Item = (), Error = Expired> {
+///         self.timer.timeout(self.delay)
+///     }
+/// }
+/// 
+/// impl<S> Middleware<S> for Timeout
+///     where S: Service,
+///           S::Error: From<Expired>,
+/// {
+///     type WrappedService = TimeoutService<S>;
+///     
+///     fn wrap(self, upstream: S) -> TimeoutService<S> {
+///         TimeoutService { timeout: self, upstream }
+///     }
+/// }
+///
+///
+/// // This service implements the Timeout behavior.
+/// pub struct TimeoutService<S> {
+///     upstream: S,
+///     timeout: Timeout,
+/// }
+///
+/// impl<S> Service for TimeoutService<S>
+///     where S: Service,
+///           S::Error: From<Expired>,
+/// {
+///     type Request = S::Request;
+///     type Response = S::Response;
+///     type Error = S::Error;
+///     type Future = Box<Future<Item = Self::Response, Error = Self::Error>>;
+///
+///     fn call(&self, req: Self::Req) -> Self::Future {
+///         let timeout = self.timeout.timeout()
+///             .and_then(|timeout| Err(Self::Error::from(timeout)));
+///
+///         self.upstream.call(req)
+///             .select(timeout)
+///             .map(|(v, _)| v)
+///             .map_err(|(e, _)| e)
+///             .boxed()
+///     }
+/// }
+///
+/// ```
+///
+/// The above timeout implementation is decoupled from the underlying protocol
+/// and is also decoupled from client or server concerns. In other words, the
+/// same timeout middleware could be used in either a client or a server.
+pub trait Middleware<S: Service> {
+    /// The service produced by wrapping this middleware around another
+    /// service.
+    type WrappedService: Service;
```

Comment:
If we are going to bike shed, I think it could be nice to come up w/ a shorter name for this associated type... I would say `Upstream`, but `up` and `down` seem to get everyone confused...

Reply:
How about `next`. In a pipeline sense, the "wrapped" service is the next called in the pipeline. `next` is commonly used by other language frameworks that have a similar mechanism for middleware.